src/walkUp.js

```diff
@@ -0,0 +1,19 @@
+import { getParent } from "./parents.js";
+import { assertParentPointers } from "./util.js";
+
+/**
+ * Walk up the tree from the given node, calling the callback on each node.
+ * @param {object} node
+ * @param {function} callback
+ * @returns {any} The first non-undefined value returned by the callback, otherwise undefined
+ */
+export default function walkUp (node, callback) {
+	while (node) {
+		const ret = callback(node);
```

Comment:
Currently even in a tree with zero parent pointers you get the callback executed once. Is that what's best for the use cases that use this? Not a rhetorical question, I don't know. We need to look at the code using this function to see what would work well for these use cases.

Reply:
The main issue I see is the fact that we could execute an arbitrary number of callbacks before running into an error in the case where some ancestor node `n` nodes away doesn't have parent pointers set. The best option might even be to walk up the entire tree before executing any callbacks, and asserting parent pointers, and then after completing the first walk up, doing a second pass where you execute the callback. But of course, this also has a performance increase factor of 2, but might be worth it! What do you think @LeaVerou